File: windows/src/LanStash.Domain/DesktopCloudDrivePlanning.cs
namespace LanStash.Domain;

public sealed record DesktopDrivePlannedFile(
    string RemotePath,
    long SizeBytes,
    DateTimeOffset? ModifiedAt);

public enum DesktopDrivePlanIssueKind
{
    InaccessibleFolder,
    UnknownFileSize,
    InvalidPath,
    ItemLimitReached,
    SizeOverflow,
    Cancelled,
}

public sealed record DesktopDrivePlanIssue(
    DesktopDrivePlanIssueKind Kind,
    string? RemotePath = null);

public sealed record DesktopDriveCachePlan(
    IReadOnlyList<DesktopDrivePlannedFile> Files,
    IReadOnlyList<string> Folders,
    IReadOnlyList<DesktopDrivePlanIssue> Issues,
    long TotalBytes,
    long LargestFileBytes,
    int FolderCount)
{
    public bool IsComplete => Issues.Count == 0;
}

public sealed record DesktopDrivePlanningProgress(
    int FolderCount,
    int FileCount,
    long DiscoveredBytes);

public enum DesktopDriveOfflinePhase
{
    Planning,
    CheckingSpace,
    Preparing,
    Downloading,
    Completed,
    Cancelled,
    Failed,
}

public sealed record DesktopDriveOfflineProgress(
    DesktopDriveOfflinePhase Phase,
    int CompletedFiles = 0,
    int TotalFiles = 0,
    long CompletedBytes = 0,
    long TotalBytes = 0,
    long? RequiredBytes = null,
    long? AvailableBytes = null,
    long? ShortageBytes = null,
    string? VolumeName = null);

public static class DesktopDriveTreePlanner
{
    public static async Task<DesktopDriveCachePlan> BuildAsync(
        IReadOnlyList<string> rootFolders,
        Func<string, int, int, CancellationToken, Task<FilePage>> loadPage,
        IReadOnlyList<DesktopDrivePlannedFile>? rootFiles = null,
        int itemLimit = 1_000_000,
        int pageSize = 500,
        IProgress<DesktopDrivePlanningProgress>? progress = null,
        CancellationToken cancellationToken = default)
    {
        if (itemLimit <= 0 || pageSize <= 0)
        {
            return Result(
                [],
                [],
                [new(DesktopDrivePlanIssueKind.ItemLimitReached)],
                0,
                0,
                0);
        }

        var queue = new List<string>(rootFolders);
        var nextFolderIndex = 0;
        var visitedFolders = new HashSet<string>(StringComparer.Ordinal);
        var folders = new List<string>();
        var visitedFiles = new HashSet<string>(StringComparer.Ordinal);
        var files = new List<DesktopDrivePlannedFile>();
        var issues = new List<DesktopDrivePlanIssue>();
        long totalBytes = 0;
        long largestFileBytes = 0;

        foreach (var file in rootFiles ?? [])
        {
            if (files.Count >= itemLimit)
            {
                issues.Add(new(DesktopDrivePlanIssueKind.ItemLimitReached));
                break;
            }
            var path = DesktopDrivePath.Normalize(file.RemotePath);
            if (path is null)
            {
                issues.Add(new(
                    DesktopDrivePlanIssueKind.InvalidPath,
                    file.RemotePath));
                continue;
            }
            if (file.SizeBytes < 0)
            {
                issues.Add(new(
                    DesktopDrivePlanIssueKind.UnknownFileSize,
                    path));
                continue;
            }
            if (!visitedFiles.Add(path))
            {
                continue;
            }
            try
            {
                totalBytes = checked(totalBytes + file.SizeBytes);
            }
            catch (OverflowException)
            {
                issues.Add(new(DesktopDrivePlanIssueKind.SizeOverflow));
                totalBytes = long.MaxValue;
                largestFileBytes = Math.Max(
                    largestFileBytes,
                    file.SizeBytes);
                break;
            }
            largestFileBytes = Math.Max(largestFileBytes, file.SizeBytes);
            files.Add(file with { RemotePath = path });
        }
        progress?.Report(new(0, files.Count, totalBytes));

        while (nextFolderIndex < queue.Count)
        {
            if (cancellationToken.IsCancellationRequested)
            {
                issues.Add(new(DesktopDrivePlanIssueKind.Cancelled));
                break;
            }
            var rawFolder = queue[nextFolderIndex++];
            var folder = DesktopDrivePath.Normalize(rawFolder);
            if (folder is null)
            {
                issues.Add(new(DesktopDrivePlanIssueKind.InvalidPath, rawFolder));
                continue;
            }
            if (!visitedFolders.Add(folder))
            {
                continue;
            }
            folders.Add(folder);

            var offset = 0;
            var folderFailed = false;
            do
            {
                try
                {
                    cancellationToken.ThrowIfCancellationRequested();
                    var page = await loadPage(
                        folder,
                        offset,
                        pageSize,
                        cancellationToken).ConfigureAwait(false);
                    foreach (var item in page.Items)
                    {
                        if (visitedFolders.Count + visitedFiles.Count >= itemLimit)
                        {
                            issues.Add(new(DesktopDrivePlanIssueKind.ItemLimitReached));
                            return Result(
                                files,
                                folders,
                                issues,
                                totalBytes,
                                largestFileBytes,
                                visitedFolders.Count);
                        }
                        if (item.IsDirectory)
                        {
                            queue.Add(item.Path);
                            continue;
                        }
                        var normalizedPath = DesktopDrivePath.Normalize(item.Path);
                        if (normalizedPath is null)
                        {
                            issues.Add(new(
                                DesktopDrivePlanIssueKind.InvalidPath,
                                item.Path));
                            continue;
                        }
                        if (!visitedFiles.Add(normalizedPath))
                        {
                            continue;
                        }
                        if (item.Size < 0)
                        {
                            issues.Add(new(
                                DesktopDrivePlanIssueKind.UnknownFileSize,
                                normalizedPath));
                            continue;
                        }
                        try
                        {
                            totalBytes = checked(totalBytes + item.Size);
                        }
                        catch (OverflowException)
                        {
                            issues.Add(new(DesktopDrivePlanIssueKind.SizeOverflow));
                            return Result(
                                files,
                                folders,
                                issues,
                                long.MaxValue,
                                Math.Max(largestFileBytes, item.Size),
                                visitedFolders.Count);
                        }
                        largestFileBytes = Math.Max(largestFileBytes, item.Size);
                        files.Add(new(
                            normalizedPath,
                            item.Size,
                            item.ModifiedAt));
                    }
                    progress?.Report(new(
                        visitedFolders.Count,
                        files.Count,
                        totalBytes));
                    var nextOffset = page.Offset + page.Items.Count;
                    if (nextOffset <= offset ||
                        page.Items.Count == 0 ||
                        nextOffset >= page.Total)
                    {
                        break;
                    }
                    offset = nextOffset;
                }
                catch (OperationCanceledException)
                {
                    issues.Add(new(DesktopDrivePlanIssueKind.Cancelled));
                    return Result(
                        files,
                        folders,
                        issues,
                        totalBytes,
                        largestFileBytes,
                        visitedFolders.Count);
                }
                catch
                {
                    issues.Add(new(
                        DesktopDrivePlanIssueKind.InaccessibleFolder,
                        folder));
                    folderFailed = true;
                }
            }
            while (!folderFailed);
        }

        return Result(
            files,
            folders,
            issues,
            totalBytes,
            largestFileBytes,
            visitedFolders.Count);
    }

    private static DesktopDriveCachePlan Result(
        IReadOnlyList<DesktopDrivePlannedFile> files,
        IReadOnlyList<string> folders,
        IReadOnlyList<DesktopDrivePlanIssue> issues,
        long totalBytes,
        long largestFileBytes,
        int folderCount) =>
        new(files, folders, issues, totalBytes, largestFileBytes, folderCount);
}

File: apple/Packages/DsmCore/Sources/DesktopCloudDrivePlanning.swift
import Foundation

public struct DesktopDrivePlannedFile: Equatable, Sendable {
    public let remotePath: String
    public let sizeBytes: Int64
    public let modifiedAt: Date?

    public init(
        remotePath: String,
        sizeBytes: Int64,
        modifiedAt: Date?
    ) {
        self.remotePath = remotePath
        self.sizeBytes = sizeBytes
        self.modifiedAt = modifiedAt
    }
}

public enum DesktopDrivePlanIssueKind: String, Codable, Hashable, Sendable {
    case inaccessibleFolder
    case unknownFileSize
    case invalidPath
    case itemLimitReached
    case sizeOverflow
    case cancelled
}

public struct DesktopDrivePlanIssue: Equatable, Sendable {
    public let kind: DesktopDrivePlanIssueKind
    public let remotePath: String?

    public init(kind: DesktopDrivePlanIssueKind, remotePath: String? = nil) {
        self.kind = kind
        self.remotePath = remotePath
    }
}

public struct DesktopDriveCachePlan: Equatable, Sendable {
    public let files: [DesktopDrivePlannedFile]
    public let issues: [DesktopDrivePlanIssue]
    public let totalBytes: Int64
    public let largestFileBytes: Int64
    public let folderCount: Int

    public init(
        files: [DesktopDrivePlannedFile],
        issues: [DesktopDrivePlanIssue],
        totalBytes: Int64,
        largestFileBytes: Int64,
        folderCount: Int
    ) {
        self.files = files
        self.issues = issues
        self.totalBytes = totalBytes
        self.largestFileBytes = largestFileBytes
        self.folderCount = folderCount
    }

    public var isComplete: Bool { issues.isEmpty }
}

public struct DesktopDrivePlanningProgress: Equatable, Sendable {
    public let folderCount: Int
    public let fileCount: Int
    public let discoveredBytes: Int64

    public init(
        folderCount: Int,
        fileCount: Int,
        discoveredBytes: Int64
    ) {
        self.folderCount = folderCount
        self.fileCount = fileCount
        self.discoveredBytes = discoveredBytes
    }
}

public enum DesktopDriveTreePlanner {
    public typealias PageLoader = @Sendable (
        _ folderPath: String,
        _ offset: Int,
        _ limit: Int
    ) async throws -> FilePage

    public static func build(
        rootFolders: [String],
        rootFiles: [DesktopDrivePlannedFile] = [],
        itemLimit: Int = 1_000_000,
        pageSize: Int = 500,
        loadPage: @escaping PageLoader,
        progress: @escaping @Sendable (DesktopDrivePlanningProgress) -> Void = { _ in }
    ) async -> DesktopDriveCachePlan {
        guard itemLimit > 0, pageSize > 0 else {
            return DesktopDriveCachePlan(
                files: [],
                issues: [.init(kind: .itemLimitReached)],
                totalBytes: 0,
                largestFileBytes: 0,
                folderCount: 0
            )
        }

        var queue = rootFolders
        var nextFolderIndex = 0
        var visited = Set<String>()
        var visitedFiles = Set<String>()
        var files: [DesktopDrivePlannedFile] = []
        var issues: [DesktopDrivePlanIssue] = []
        var totalBytes: Int64 = 0
        var largestFileBytes: Int64 = 0

        for file in rootFiles {
            guard files.count < itemLimit else {
                issues.append(.init(kind: .itemLimitReached))
                break
            }
            guard let path = DesktopDrivePath.normalized(file.remotePath) else {
                issues.append(.init(kind: .invalidPath, remotePath: file.remotePath))
                continue
            }
            guard file.sizeBytes >= 0 else {
                issues.append(.init(kind: .unknownFileSize, remotePath: path))
                continue
            }
            guard visitedFiles.insert(path).inserted else { continue }
            let sum = totalBytes.addingReportingOverflow(file.sizeBytes)
            guard !sum.overflow else {
                issues.append(.init(kind: .sizeOverflow))
                totalBytes = .max
                largestFileBytes = max(largestFileBytes, file.sizeBytes)
                break
            }
            totalBytes = sum.partialValue
            largestFileBytes = max(largestFileBytes, file.sizeBytes)
            files.append(
                DesktopDrivePlannedFile(
                    remotePath: path,
                    sizeBytes: file.sizeBytes,
                    modifiedAt: file.modifiedAt
                )
            )
        }
        progress(
            DesktopDrivePlanningProgress(
                folderCount: 0,
                fileCount: files.count,
                discoveredBytes: totalBytes
            )
        )

        while nextFolderIndex < queue.count {
            do {
                try Task.checkCancellation()
            } catch {
                issues.append(.init(kind: .cancelled))
                break
            }
            let rawFolder = queue[nextFolderIndex]
            nextFolderIndex += 1
            guard let folder = DesktopDrivePath.normalized(rawFolder) else {
                issues.append(.init(kind: .invalidPath, remotePath: rawFolder))
                continue
            }
            guard visited.insert(folder).inserted else { continue }
            var offset = 0
            var folderFailed = false

            repeat {
                do {
                    try Task.checkCancellation()
                    let page = try await loadPage(folder, offset, pageSize)
                    for item in page.items {
                        if visited.count + files.count >= itemLimit {
                            issues.append(.init(kind: .itemLimitReached))
                            return result(
                                files: files,
                                issues: issues,
                                totalBytes: totalBytes,
                                largestFileBytes: largestFileBytes,
                                folderCount: visited.count
                            )
                        }
                        if item.isDirectory {
                            queue.append(item.path)
                            continue
                        }
                        guard let normalizedPath = DesktopDrivePath.normalized(item.path) else {
                            issues.append(
                                .init(kind: .invalidPath, remotePath: item.path)
                            )
                            continue
                        }
                        guard visitedFiles.insert(normalizedPath).inserted else {
                            continue
                        }
                        guard let size = item.sizeBytes, size >= 0 else {
                            issues.append(
                                .init(kind: .unknownFileSize, remotePath: normalizedPath)
                            )
                            continue
                        }
                        let sum = totalBytes.addingReportingOverflow(size)
                        guard !sum.overflow else {
                            issues.append(.init(kind: .sizeOverflow))
                            return result(
                                files: files,
                                issues: issues,
                                totalBytes: .max,
                                largestFileBytes: max(largestFileBytes, size),
                                folderCount: visited.count
                            )
                        }
                        totalBytes = sum.partialValue
                        largestFileBytes = max(largestFileBytes, size)
                        files.append(
                            DesktopDrivePlannedFile(
                                remotePath: normalizedPath,
                                sizeBytes: size,
                                modifiedAt: item.times?.modifiedAt
                            )
                        )
                    }
                    progress(
                        DesktopDrivePlanningProgress(
                            folderCount: visited.count,
                            fileCount: files.count,
                            discoveredBytes: totalBytes
                        )
                    )
                    let nextOffset = page.offset + page.items.count
                    if !page.hasMore || nextOffset <= offset || page.items.isEmpty {
                        break
                    }
                    offset = nextOffset
                } catch is CancellationError {
                    issues.append(.init(kind: .cancelled))
                    return result(
                        files: files,
                        issues: issues,
                        totalBytes: totalBytes,
                        largestFileBytes: largestFileBytes,
                        folderCount: visited.count
                    )
                } catch {
                    issues.append(
                        .init(kind: .inaccessibleFolder, remotePath: folder)
                    )
                    folderFailed = true
                }
            } while !folderFailed
        }

        return result(
            files: files,
            issues: issues,
            totalBytes: totalBytes,
            largestFileBytes: largestFileBytes,
            folderCount: visited.count
        )
    }

    private static func result(
        files: [DesktopDrivePlannedFile],
        issues: [DesktopDrivePlanIssue],
        totalBytes: Int64,
        largestFileBytes: Int64,
        folderCount: Int
    ) -> DesktopDriveCachePlan {
        DesktopDriveCachePlan(
            files: files,
            issues: issues,
            totalBytes: totalBytes,
            largestFileBytes: largestFileBytes,
            folderCount: folderCount
        )
    }
}
